### ISSUE TITLE
BUG: strings of different capacity compare not-equal

### ISSUE THREAD

**Issue by jzwinck** [numpy/numpy#8982]
First, some setup code:

    import numpy as np
    text = np.array([('hello',), ('world',)], [('x', 'S7')])
    good = np.array([('world',)], [('x', 'S7')])
    long = np.array([('world',)], [('x', 'S8')])

Now, `np.in1d(text, good)` works (giving `[False, True]`), but `np.in1d(text, long)` prints a warning and gives the wrong result:

    arraysetops.py:379: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
      mask |= (ar1 == a)
    Out: array([False, False], dtype=bool)

However, `np.in1d(text['x'], long['x'])` works.

The error seems to stem from the fact that `text[1] == long[0]` does not work:

    ipython:1: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
    Out: False

But again, `text[1]['x'] == long[0]['x']` does work.

The fact that the maximum storage capacity for two strings is different should not change the result of their comparison.  I would expect all the above code to work without warnings.

Tested using NumPy 1.11.1.

**Comment by eric-wieser** [comment]
Note that `np.equal(long, good)` also fails with a slightly different error to `long == good`.

There's a detailed comment about it [here](https://github.com/numpy/numpy/blob/dcafaf4983214a0c1ceefc154f1dcdb3d04db748/numpy/core/src/umath/ufunc_object.c#L875-L959)

**Comment by badock** [comment]
I took a look to this issue and noticed that it may come from the [**array_richcompare**](https://github.com/numpy/numpy/blob/77ff3d69dfc1d7f2410ff373900c467d829863b2/numpy/core/src/multiarray/arrayobject.c#L1300) function which compares two arrays.

In the ```text[1] == long[0]``` expression, ```text``` and ```long``` are two arrays which stores strings of different sizes (respectively 7 bytes and 8 bytes). The **array_richcompare** function calls the [**PyArray_CanCastTypeTo**](https://github.com/numpy/numpy/blob/77ff3d69dfc1d7f2410ff373900c467d829863b2/numpy/core/src/multiarray/convert_datatype.c#L678) function.

**PyArray_CanCastTypeTo** takes two arrays and a ```casting``` parameters. [Here](https://github.com/numpy/numpy/blob/77ff3d69dfc1d7f2410ff373900c467d829863b2/numpy/core/src/multiarray/arrayobject.c#L1384), the ```casting``` parameter is set to [*NPY_EQUIV_CASTING*](https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/ndarraytypes.h#L195), which causes **PyArray_CanCastTypeTo** to compare the types of elements stored in the two given array, and to [strictly check](https://github.com/numpy/numpy/blob/77ff3d69dfc1d7f2410ff373900c467d829863b2/numpy/core/src/multiarray/convert_datatype.c#L786) that these types have the same size (this is not the case in the ```text[1] == long[0]``` expression).

A solution to fix this issue can be to modify the way **array_richcompare** calls **PyArray_CanCastTypeTo**, by using a different value for the ```casting``` parameter. For example using [*NPY_SAFE_CASTING*](https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/ndarraytypes.h#L197) or [*NPY_SAME_KIND_CASTING*](https://github.com/numpy/numpy/blob/master/numpy/core/include/numpy/ndarraytypes.h#L199) solves this issue and does not cause any unit test to fail.

**Comment by rainwoodman** [comment]
Why is the code here:
https://github.com/numpy/numpy/blob/77ff3d69dfc1d7f2410ff373900c467d829863b2/numpy/core/src/multiarray/arrayobject.c#L1310

Not called? It claims the section is dealing with 'strings'?

**Comment by badock** [comment]
I think that the code is not called because arrays are not considered as being "arrays of string", but rather as being "arrays of void". It seems that during the creation of an array, providing a "tuple value" for the *dtype* argument results in the array becoming an "array of void":

``` python
explicitly_typed_array = np.array([('world',)], [('x', 'S7')])
print(explicitly_typed_array.dtype.type) # <type 'numpy.void'>
implicitly_typed_array = np.array([('world',)])
print(implicitly_typed_array.dtype.type) # <type 'numpy.string_'>
```

**Comment by rainwoodman** [comment]
I see. The structured array comparison is a byte-by-byte comparison. The ordering is not defined if the two elements are of different length.

**Comment by badock** [comment]
Do you see a way to address this issue?

**Comment by eric-wieser** [comment]
> The structured array comparison is a byte-by-byte comparison. 

I don't think this is true. It should be field-by-field, but I think that the special casing for string types is missing there.

The best fix here is to allow ufuncs to work with flexible types (strings + void). Unfortunately, this is hard.

**Comment by jzwinck** [comment]
@eric-wieser you said:
> The best fix here is to allow ufuncs to work with flexible types (strings + void). Unfortunately, this is hard.

Has any progress been made in this area since 2017?  If it's still going to be hard, do you think it's worth doing at all?  If so, can you explain perhaps a first step which will get us closer--something someone can bite off and chew without having the entire hard problem to deal with all at once?

**Comment by seberg** [comment]
We could try to patch things over in the `==` code of void probably. Adding a ufunc is generally possible, although there are currently some subtleties that may need a lot of care.  Since I should get to revamping how ufuncs work here in the next months, I am not very enthusiastic about attempting that.

But, if the issue is currently mainly in `PyArray_CanCastTypeTo` then, I think it can very much be fixed of ufuncs. That part of code should also move around, but such bugfixes are easy to keep around. (There is one open PR about structured dtype type promotion and maybe can-cast though).

**Comment by rainwoodman** [comment]
S Strings are not necessary 0 terminated. So "world\0" and "world\0\0" are actually different bytes in that sense of equal.

**Comment by seberg** [comment]
@rainwoodman not sure I understand the point. The string dtype ignores any zero padding, so based on string equality, they are equal.  And structured equality should use structure-element equality internally (and does).

This is clearly a bug and surprising behaviour. I can see that we may need a release note when fixing it, but I do not see we can argue that they are different.  If you want a different result on that level, the user has to cast to `"V7"` or similar.

**Comment by rainwoodman** [comment]
Thanks for the clarification. I checked the doc and yes S indeed is for null-terminated bytes-string -- thus it shall not compare beyond the zero terminator. I was confused.
